torii/sim/wasm/_wasmrtl.py

```diff
@@ -0,0 +1,585 @@
+			emitter.append(f'(local.set $next_{signal_index} (global.get $slots_{signal_index}_next))')
+		compiler = cls(state, emitter)
+		compiler(stmt)
+		for signal_index in output_indexes:
+			emitter.append(f'(call $slots_{signal_index}_set (local.get $next_{signal_index}))')
+
+		# TODO: only import signals that are actually needed. They need to be in order!
+		for slot in state.slots:
+			signal_index = state.get_signal(slot.signal)
+			emitter.add_global_var(f'slots_{signal_index}_curr')
+			emitter.add_global_var(f'slots_{signal_index}_next')
+			emitter.add_func(f'slots_{signal_index}_set')
+
+		output_code = emitter.flush()
+		return output_code
+
+class _WASMFragmentCompiler:
+	def __init__(self, state) -> None:
+		self.state = state
+
+	def __call__(self, fragment: Fragment):
+		processes = set()
+
+		for domain_name, domain_signals in fragment.drivers.items():
+			domain_stmts = LHSGroupFilter(domain_signals)(fragment.statements)
+			domain_process = WASMRTLProcess(is_comb = domain_name is None)
+
+			emitter = _WASMEmitter()
+			if domain_name is None:
+				for signal in domain_signals:
+					signal_index = self.state.get_signal(signal)
+					emitter.add_variable(f'next_{signal_index}')
+					emitter.append(f'(local.set $next_{signal_index} (i64.const {signal.reset}))')
+
+				inputs = SignalSet()
+				_StatementCompiler(self.state, emitter, inputs = inputs)(domain_stmts)
+
+				for input in inputs:
+					self.state.add_trigger(domain_process, input)
+
+			else:
+				domain = fragment.domains[domain_name]
+				clk_trigger = 1 if domain.clk_edge == 'pos' else 0
+				self.state.add_trigger(domain_process, domain.clk, trigger = clk_trigger)
+				if domain.rst is not None and domain.async_reset:
+					rst_trigger = 1
+					self.state.add_trigger(domain_process, domain.rst, trigger = rst_trigger)
+
+				for signal in domain_signals:
+					signal_index = self.state.get_signal(signal)
+					emitter.add_variable(f'next_{signal_index}')
+					emitter.append(f'(local.set $next_{signal_index} (global.get $slots_{signal_index}_next))')
+
+				_StatementCompiler(self.state, emitter)(domain_stmts)
+
+			for signal in domain_signals:
+				signal_index = self.state.get_signal(signal)
+				emitter.append(f'(call $slots_{signal_index}_set (local.get $next_{signal_index}))')
+
+			glob_vars = []
+			glob_func = []
+
+			for slot in self.state.slots:
+				signal_index = self.state.get_signal(slot.signal)
+				emitter.add_global_var(f'slots_{signal_index}_curr')
+				emitter.add_global_var(f'slots_{signal_index}_next')
+				emitter.add_func(f'slots_{signal_index}_set')
+				glob_func.append(Func(self.state.store, FuncType([ValType.i64()], []), slot.set))
+				# current is first, next is second
+				glob_vars.append(slot.curr.wasm_value())
+				glob_vars.append(slot.next.wasm_value())
+
+			# Make sure that globals are imported in the order they are imported
+			# in the WASMEmitter
+			glob_vars.extend(glob_func)
+			module_code = emitter.flush()
```

Comment:
Have the ability to dump and introspect the generated WASM like in `pyrtl` with the `getenv('TORII_PYSIM_DUMP')` would be handy.